pii_recognition/registers/registry.py

```diff
@@ -1,17 +1,14 @@
-from abc import ABCMeta, abstractmethod
-from typing import TypeVar
+from typing import Generic, Optional, Type, TypeVar
 
-T = TypeVar("T")
+T_co = TypeVar("T_co", covariant=True)
 
 
-class Registry(dict, metaclass=ABCMeta):
-    def __init__(self):
-        self.add_predefines()
+class Registry(dict, Generic[T_co]):
```

Comment:
so basically here you are trying to make sure that the items registered are immutable? hence use the covariance type?

Reply:
To my understanding there is no strong link between immutable and covariant, we can have immutable to be invariance [here](https://stackoverflow.com/questions/676615/why-is-scalas-immutable-set-not-covariant-in-its-type). 

[This](https://mypy.readthedocs.io/en/latest/generics.html#variance-of-generic-types) is a definition from mypy on covariant, basically saying if `CRFRecogniser` is a subtype of `Recogniser`, then `Registry[CRFRecogniser]` is a subtype of `Registry[Recogniser]`.

Reply:
I found this example maybe useful for understanding the concepts. Covariant, contravariant and invariant are indeed very confusing concepts.
```
from typing import Generic, Type, TypeVar

T_co = TypeVar('T_co', covariant=True)

class Tokeniser: pass
class SpaceSplit(Tokeniser): pass


class Registry(Generic[T_co]):
    def __init__(self, item: Type[T_co]) -> None:
        self._item = item

    def get_content(self) -> T_co:  # my would complain here for contravariant type
        return self._item()

def look_into_registry(registry: Registry[Tokeniser]): ...

my_tokeniser = Registry(SpaceSplit)
look_into_registry(my_tokeniser)  # mypy would complain here for an invariant type

```